gossip/src/crds_value.rs

```diff
@@ -490,84 +496,193 @@ impl Sanitize for NodeInstance {
+    encoded: Vec<U16>,
+}
+
+impl RunLengthEncoding {
+    // Return whether the encoded vec is already full.
+    fn valid_and_has_space(total_serde_bytes: &mut usize, count: usize) -> bool {
+        match u16::try_from(count) {
+            Ok(current_bit_count) => {
+                let serde_bits = u16::BITS - current_bit_count.leading_zeros();
+                let serde_bytes = ((serde_bits + 6) / 7).max(1) as usize;
+                if *total_serde_bytes + serde_bytes > RestartLastVotedForkSlots::MAX_SPACE {
+                    false
+                } else {
+                    *total_serde_bytes += serde_bytes;
+                    true
+                }
+            }
+            Err(e) => {
+                info!("Failed to convert count to u16: {e:?}");
+                false
+            }
+        }
+    }
+
+    pub fn new(bits: &BitVec<u8>) -> Self {
+        let mut total_serde_bytes = 0;
+        let encoded = (0..bits.len())
+            .map(|i| bits.get(i))
+            .dedup_with_count()
+            .map(|(count, _)| count)
+            .take_while(|count| Self::valid_and_has_space(&mut total_serde_bytes, *count))
+            .map(|count| U16(count.try_into().unwrap()))
+            .collect();
+        Self { encoded }
+    }
+
+    pub fn slots_count(&self) -> usize {
+        self.encoded.iter().map(|x| x.0 as usize).sum::<usize>()
+    }
+
+    pub fn to_slots(&self, last_slot: Slot, min_slot: Slot) -> Vec<Slot> {
+        let mut result: Vec<Slot> = self
+            .encoded
+            .iter()
+            .zip([1, 0].iter().cycle())
+            .flat_map(|(bit_count, bit)| std::iter::repeat(bit).take(bit_count.0.into()))
+            .enumerate()
+            .filter_map(|(offset, bit): (usize, &i32)| {
+                if *bit > 0 {
+                    Some(last_slot.saturating_sub(offset as Slot))
+                } else {
+                    None
+                }
+            })
+            .take(MAX_SLOTS_PER_ENTRY)
+            .take_while(|slot| *slot >= min_slot)
+            .collect();
+        result.reverse();
+        result
+    }
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq, AbiExample)]
+pub struct RawOffsets {
+    bits: BitVec<u8>,
+}
+
+impl RawOffsets {
+    pub fn new(mut bits: BitVec<u8>) -> Self {
+        bits.truncate(RestartLastVotedForkSlots::MAX_SPACE as u64 * 8);
+        bits.shrink_to_fit();
+        Self { bits }
+    }
+
+    pub fn to_slots(&self, last_slot: Slot, min_slot: Slot) -> Vec<Slot> {
+        let mut result = Vec::new();
+        for i in 0..self.bits.len() {
+            if self.bits.get(i) {
+                let slot = last_slot.saturating_sub(i as Slot);
+                if slot < min_slot {
+                    return result;
+                }
+                result.push(slot);
+                if result.len() > MAX_SLOTS_PER_ENTRY {
+                    break;
+                }
+            }
+        }
```

Comment:
This can also be written using iterators instead of manual `for` loops and `if` branches;
something like:
```rust
(0..self.bits.len())
  .filter(|k| self.bits.get(k))
  .map(|offset| last_slot.saturing_sub(offset as Slot))
  .take_while(|slot| slot >= min_slot)
  .take(MAX_SLOTS_PER_ENTRY)
```

Reply:
Done.